### DISCUSSION TITLE
Duplicated stylesheets when manually compiling with Babel

### DISCUSSION THREAD

**Discussion by jonathanj** [nativewind/nativewind#230] [General]
I am experimenting with NativeWind in a React Native project, which is distributed as a library. I wanted to look at the generated output:

```
yarn babel src --out-dir tmp --config-file ./babel.config.js --extensions .ts,.tsx
```

I noticed that at the bottom of every compiled file there is a `NativeWindStyleSheet.create` call for all the styles used in the file, redefined again in every file, which is going to add up over files and number of styles used.

Is it possible to have these automatically extracted to a common stylesheet instead of redefined repeatedly?

I followed the [Tailwind CLI](https://www.nativewind.dev/guides/cli-native) instructions which generated a `nativewind-output.js` , but trying to use the [Babel transform-only](https://www.nativewind.dev/guides/babel-transform-only) the full `NativeWindStyleeSheet.create` call still appears at the bottom of every source file.

Have I misunderstood something or maybe misconfigured something?

<details>

```
// babel.config.js
module.exports = {
  presets: ['module:metro-react-native-babel-preset'],
  plugins: [
    ['nativewind/babel', {mode: 'transformOnly'}],
  ],
};

$ yarn babel --version
7.17.6 (@babel/core 7.17.7)

$ yarn why nativewind
├─ <snip> [0ae88]
│  └─ nativewind@npm:2.0.8 [fd890] (via npm:^2.0.8 [fd890])

// tailwind.config.js
/** @type {import('tailwindcss').Config} */
const nativewind = require("nativewind/tailwind/native");

module.exports = {
  content: ['./src/**/*.{js,jsx,ts,tsx}', './_intro/**/*.{js,jsx,ts,tsx}'],
  theme: {
    extend: {},
  },
  plugins: [nativewind()],
}
```

</details>

**Comment by marklawlor** [comment]
> I am experimenting with NativeWind in a React Native project, which is distributed as a library

If you are expecting your users to be NativeWind users it might be easier to distribute your styles via a Tailwind plugin. Your plugin could either append the path to your components to the users `content` or you could just append all the classes you use to `safeList`. The advantage of this system if that you'll never run into peerDependancy issues.

If you'd like your library to be used by non-nativewind users you should use either the [PostCSS cli](https://www.nativewind.dev/guides/postcss) or the [Tailwind CLI](https://www.nativewind.dev/guides/cli-native) to generate a JS file with just your styles. These styles will not be duplicated.

> I noticed that at the bottom of every compiled file there is a NativeWindStyleSheet.create call for all the styles used in the file, redefined again in every file, which is going to add up over files and number of styles used.

This is a temporary solution for fast-reload to work in local development, but it increases bundle size in production. I know of a better way to do this I just haven't had the time to implement it yet and didn't want to block 2.0. Fix should be coming "soon"

> I followed the [Tailwind CLI](https://www.nativewind.dev/guides/cli-native) instructions which generated a nativewind-output.js , but trying to use the [Babel transform-only](https://www.nativewind.dev/guides/babel-transform-only) the full NativeWindStyleeSheet.create call still appears at the bottom of every source file.

Sounds like a bug. Edit: Fixed in `v2.0.9`

**Reply by jonathanj** [reply]
Thanks so much for the feedback, and for fixing the transform-only issue!

If you don't mind sharing, I'm curious what the (high-level) plan for the better way is.

**Reply by marklawlor** [reply]
> If you don't mind sharing, I'm curious what the (high-level) plan for the better way is.

Writing the styles to `node_modules/.cache/nativewind` as a single file.

When I wrote the library, RNW 0.18 hadn't been released so I was using native styles for both native & web. Most web frameworks by default are not configured to hot-reload on changed to node_modules, and I didn't want this weird extra file sitting around. So I wrote the styles into the file that just got saved. Things have changed now with RNW 0.18 and web projects should be using CSS which doesn't require the compiled styles, so I can do away with this workaround.

RN projects watch the node_modules by default (side effect of transpiling them) so I should be able to write the new styles to to a cache folder. This is similar to how tools like `@prisma/client` work

That's the theory anyway, I need to do some testing :smile: But it will fix a couple of issues with NativeWind and will speed up your project builds

**Comment by danstepanov** [comment]
Closing this discussion, if this is still relevant, feel free to open a new one.
